File: Programming/ecobike_rental/lib/model/service/network/response/bike.dart
class BikeResponse {
  int bikeId;
  String bikeName;
  String description;
  int startingRent;
  int hourlyRent;
  int type;
  String licensePlates;
  int batterCapacity;
  double powerDrain;
  int deposit;
  String rentals;

  BikeResponse(
      {this.bikeId,
      this.bikeName,
      this.description,
      this.startingRent,
      this.hourlyRent,
      this.type,
      this.licensePlates,
      this.batterCapacity,
      this.powerDrain,
      this.deposit,
      this.rentals});

  BikeResponse.fromJson(Map<String, dynamic> json) {
    bikeId = json['bikeId'];
    bikeName = json['bikeName'];
    description = json['description'];
    startingRent = json['startingRent'];
    hourlyRent = json['hourlyRent'];
    type = json['type'];
    licensePlates = json['licensePlates'];
    batterCapacity = json['batterCapacity'];
    powerDrain = json['powerDrain'];
    deposit = json['deposit'];
    rentals = json['rentals'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['bikeId'] = this.bikeId;
    data['bikeName'] = this.bikeName;
    data['description'] = this.description;
    data['startingRent'] = this.startingRent;
    data['hourlyRent'] = this.hourlyRent;
    data['type'] = this.type;
    data['licensePlates'] = this.licensePlates;
    data['batterCapacity'] = this.batterCapacity;
    data['powerDrain'] = this.powerDrain;
    data['deposit'] = this.deposit;
    data['rentals'] = this.rentals;
    return data;
  }
}

File: Programming/ecobike_rental_server/TKXDPM_API/Model/Bike.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using AutoMapper;
using TKXDPM_API.Controllers.CalculateFee;

namespace TKXDPM_API.Model
{
    public class Bike
    {
        [Key] public int BikeId { get; set; }
        [Column(TypeName = "varchar(255)")] public string BikeName { get; set; }
        [Column(TypeName = "varchar(512)")] public string BikeImage { get; set; }
        public string Description { get; set; }
        public int Size { get; set; }
        public int StartingRent { get; set; }
        public int HourlyRent { get; set; }
        public BikeType Type { get; set; }
        [Column(TypeName = "varchar(255)")] public string LicensePlates { get; set; }
        public int BatterCapacity { get; set; }
        public float PowerDrain { get; set; }
        public int Deposit { get; set; }
        public ICalculateFee CalculateBikeFee = new CalculateFeeNormal();

        public List<Rental> Rentals { get; set; }

        public Rental FindRental()
        {
            var rentals = Rentals.Where(rental => rental.Transaction.BookedEndDateTime == DateTime.MinValue).ToArray();
            return rentals[0];
        }

        public bool IsRent()
        {
            return Rentals != null &&
                   Rentals.Any(rental =>
                       rental.Transaction != null && (
                           rental.Transaction.BookedEndDateTime == DateTime.MinValue ||
                           rental.Transaction.BookedEndDateTime >= DateTime.Now));
        }

        public bool CheckDeposit(int deposit)
        {
            var condition = new Dictionary<BikeType, int>()
            {
                {BikeType.Single, 400000},
                {BikeType.Double, 550000},
                {BikeType.Electric, 700000}
            };

            return deposit >= condition[Type];
        }

        private int CalculateBaseFee(double minutes)
        {
            double fee = 0;
            if (minutes < 10)
            {
                fee = 0;
            }
            else if (minutes <= 30)
            {
                fee = StartingRent;
            }
            else
            {
                fee = Math.Ceiling((minutes - 30) / 15) * HourlyRent + StartingRent;
            }

            return (int) fee;
        }

        public int CalculateFee(double minutes)
        {
            return CalculateBikeFee.CalculateFee(minutes, Type);
        }

        public static Bike ConvertByType(Bike bike)
        {
            return bike.Type switch
            {
                BikeType.Single => (SingleBike) bike,
                BikeType.Double => (DoubleBike) bike,
                BikeType.Electric => (ElectricBike) bike,
                _ => throw new ArgumentOutOfRangeException()
            };
        }
    }

    public class BikeResponse
    {
        public static BikeResponse CreateByMapper(Bike bike, IMapper mapper)
        {
            var bikeResponse = mapper.Map<BikeResponse>(bike);
            bikeResponse.IsRented = bike.IsRent();
            return bikeResponse;
        }

        public int BikeId { get; set; }
        public string BikeName { get; set; }
        public string BikeImage { get; set; }
        public string Description { get; set; }
        public int StartingRent { get; set; }
        public int HourlyRent { get; set; }
        public BikeType Type { get; set; }
        public string LicensePlates { get; set; }
        public int BatterCapacity { get; set; }
        public float PowerDrain { get; set; }
        public int Deposit { get; set; }
        public bool IsRented { get; set; } = false;
    }

    public enum BikeType
    {
        Single = 1,
        Double = 2,
        Electric = 3,
    }

    public struct RentFee
    {
        public int StartingRent;
        public int HourlyRent;
    }
}